tests/virt_autotest/virsh_virtual_network.pm

```diff
@@ -0,0 +1,352 @@
+
+    if (is_sle('>11')) {systemctl 'restart libvirtd';} else {script_run("service libvirtd restart");}
+
+}
+
+sub restore_guests {
+
+    my $wait_script        = "30";
+    my $get_vm_hostnames   = "virsh list  --all | grep sles | awk \'{print \$2}\'";
+    my $vm_hostnames       = script_output($get_vm_hostnames, $wait_script, type_command => 0, proceed_on_failure => 0);
+    my @vm_hostnames_array = split(/\n+/, $vm_hostnames);
+    foreach (@vm_hostnames_array)
+    {
+        script_run("virsh destroy $_");
+        script_run("virsh undefine $_");
+        script_run("virsh define $_.xml");
+	upload_logs "$_.xml";
+        script_run("rm -rf $_.xml");
+    }
+
+}
+
+sub restore_network {
+
+    my $network_mark  = "/etc/sysconfig/network/ifcfg-br0.new";
+    if (script_run("[[ -f $network_mark ]]") == 0){
+        assert_script_run("ip link set dev $virt_host_bridge down",60);
+        my $bridge_destroy = <<'EOF';
+#!/bin/bash
+ACTIVE_NET=$(ip a|awk -F': ' '/state UP/ {print $2}'|head -n1)
+interface="BOOTPROTO='dhcp'\nSTARTMODE='auto'"
+echo -e $interface >/etc/sysconfig/network/ifcfg-$ACTIVE_NET
+cat /etc/sysconfig/network/ifcfg-$ACTIVE_NET
+rm -rf /etc/sysconfig/network/ifcfg-br0*
+systemctl restart network.service;
+EOF
+         script_output($bridge_destroy, 900);
+    }
+
+}
+
+sub destroy_vir_network {
+
+    my $wait_script       = "30";
+    my $get_vnet_name     = "virsh net-list --all| tail -2| head -1| awk \'{print \$1}\'";
+    my $vnet_name         = script_output($get_vnet_name, $wait_script, type_command => 0, proceed_on_failure => 0);
+    my @vnet_name_array = split(/\n+/, $vnet_name);
+    foreach (@vnet_name_array) {script_run("virsh net-destroy $_"); script_run("rm -rf $_.xml");}
+
+    my $default_path  = "/root/libvirt_default.xml";
+    if (script_run("[[ -f $default_path ]]") == 0){assert_script_run("virsh net-define $default_path",60);assert_script_run("rm -rf $default_path");}
+
+}
+
+sub run {
+
+    #Prepare VM HOST SERVER Network Interface Configuration
+    #for libvirt virtual network testing 
+    prepare_network;
+
+    #Debug br123
+    restore_standalone;
+
+    #Remove SSH old files and Generate the key pair
+    assert_script_run "rm -rf ~/.ssh/* || true;ssh-keygen -t rsa -P '' -f ~/.ssh/id_rsa";
+ 
+    #Install required packages
+    zypper_call '-t in iptables iputils bind-utils sshpass';
+
+    #Guest SSH Setup
+    my $gi_guest = '';
+    foreach my $guest (keys %xen::guests) {
+        #Archive deployed Guests
+        assert_script_run("virsh dumpxml $guest > $guest.xml");
+        record_info "$guest", "Establishing SSH connection to $guest";
+        assert_script_run("virsh start $guest", 60);
+        save_screenshot;
+        $gi_guest = script_output("sleep 40;journalctl | grep dhcp| tail -1| grep -oE \"([0-9]{1,3}[\.]){3}[0-9]{1,3}\"");
+	#Copy the VM host SSH Key to guest systems
+        assert_script_run("sshpass -p novell ssh-copy-id -f root\@$gi_guest", 60);
```

Comment:
Is it fine to use a password in the script directly? How about connect the guest the way the guest installation test does? "$sshNoPass <guest_ip> "cat /etc/issue"

Reply:
```suggestion
use testapi;
assert_script_run("sshpass -p " . $testapi::password . " ssh-copy-id -f root\@$gi_guest", 60);
```

Reply:
Cool! Will update this code ASAP.

Reply:
In that case why don't use create a global variable for the password instead of hardcoding it everywhere?

Reply:
Code changes as below:
-        assert_script_run("sshpass -p novell ssh-copy-id -f root\@$gi_guest", 60);
+        my $get_settings_path = "/usr/share/qa/virtautolib/lib/get-settings.sh";
+       if (script_run("[[ -f $get_settings_path ]]") == 0) {
+            my $vmpasswd = script_output("$get_settings_path vm.pass");
+           assert_script_run("sshpass -p " . $vmpasswd . " ssh-copy-id -f root\@$gi_guest", 60);
+        }
+        else {
+            assert_script_run("sshpass -p " . $testapi::password . " ssh-copy-id -f root\@$gi_guest", 60);
+        }
If virt_autotest environment, try to use script get-settings.sh to define guest password as variable instead of hardcoding it everywhere. if not, also still use $testapi::password.

Reply:
This is not a solution.

Reply:
For now, I think, the current code was the solution.

Reply:
I still think you should adjust the `exec_and_insert_password();` in the way that it will use your password in your tests and the default one in the other tests. It is one `if` and the code will be much more clean and generic. cc: @alice-suse

Reply:
Great. 
I will adjust the `exec_and_insert_password();`. and then update my code here. 
Thank you so much for your comments.

Reply:
Had finished adjust the exec_and_insert_password(); please refer to the latest verification run link as below:
http://10.67.19.82/tests/265